### ISSUE TITLE
if discarding the record, dtls client how to know it session is error and restart handshake ? 

### ISSUE THREAD

**Issue by xuzhiweiand** [eclipse/californium#1413]
if discarding the record, dtls client how to know it session is error and restart handshake ? 
if dtls client do not know dtls session is empty, it can work normal.

class DTLSConnector   line 1240  to 1252 code 

			if (connectionId == null) {
				DROP_LOGGER.trace("Discarding {} records from [{}] received without existing connection",
						records.size(), peerAddress);
			} else {
				DROP_LOGGER.trace("Discarding {} records from [{},{}] received without existing connection",
						records.size(), peerAddress, connectionId);
			}

**Comment by boaks** [comment]
Basically, by the information of the layer above.
If that's CoAP and the message is a CON, then you may consider to use the timeout of that message.
See [Handling Invalid Records](https://tools.ietf.org/html/rfc6347#section-4.1.2.7).

The downside is, that it may take long. Therefore you may also tryout to trigger a "probe handshake" on retransmissions even before the message timesout all retries. But that depends on your client library.

**Comment by xuzhiweiand** [comment]
non notification，client do not know handshake is wrong. 
What should be done on the server ?  The client cannot modify behavior.

**Comment by boaks** [comment]
Yes, and if the non is dropped somewhere in the network, the client will also never know.
That's the basic decision to use NON for notifies, that you don't get a ACK and so you accept, that it may be lost.
Therefore it's recommended to uses sometimes a CON notify to ensure, that the server is still interested and receiving these notifies.

Let me add:
Using UDP comes with:
- messages may be lost (in the public internet usually about 3-5%)
- messages may be reordered (very rare, but may happen)
- messages may be spoofed (attacks with modified source addresses, hopefully rare :-) ).

DTLS only addresses that for the handshake, and for the application data it relays on the next layer to do so.
DTLS will protect you from spoofing by the MAC, if DTLS CID is not used. If DTLS CID is used, then the IETF is discussion the RRC, but for now that relays on the next layer. If that next layer is CoAP, that offers multiple "features" to deal with the three "natures" of UDP, but it's up to you to chose the right one for your requirements.

The alternative "solution", sending a message on DTLS layer back, will have also downsides.
Basically it's said, that you offer with that knowledge to attackers, which harms you more.
Currently [RFC5246 - Alert Protocol](https://tools.ietf.org/html/rfc5246#section-7.2) only defines messages, which would close the session on the other peer. Assuming spoofing, that will allow attackers to disrupt sessions easily. An extension of these alert messages mechanisms would requires a discussion in the IETF tls-mailing list. My forecast for the outcome of such a discussion is "no change" :-).
